[Enhancement] Feature/smarterbutton. Ability to navigate to a URL/Dashboard and mimic a Dummy

LuckyMallari opened the pull request:
Previous conversation:
https://github.com/openhab/org.openhab.ui.habpanel/pull/258#issuecomment-354185262

Here's the changes per suggestions:
* Changed "Navigate to dashboard" to "Navigate"
* Change Icons uitab to Display
* Added ability to navigate to an external URL
* Changed ui-select on dashboard select tag
* Use/mimic Dummy widget for Navigate type
* Added ability to show an item value

Basically if you use "Display" on a "Navigate" action type, you can display the Item name/value just like a regular Dummy widget. Same concept as previous PR: ability to navigate to a URL or a dashboard page. Use self/new tab/new window if preferred.

Demo:
![smartbutton](https://user-images.githubusercontent.com/6183345/34427044-5bc2c11c-ebf2-11e7-8005-93d4b63959d2.gif)


I formatted code based on VS Code settings.
You can see actual changes here:
https://github.com/LuckyMallari/org.openhab.ui.habpanel/commit/d56498237fe75cb67c58beb4fdbab494c5e64c21

I also removed huge blocks of code that were not used (commented out)

ghys commented on web/app/widgets/button/button.settings.tpl.html:
```diff
@@ -7,157 +7,239 @@ <h3>Button Settings</h3>
+                    </uib-tabset>
                 </div>
-                <div class="col-xs-3" style="height: 80px">
+                <div class="col-xs-3" style="height: 80px" ng-if="form.action_type !== 'navigate'">
```
An idea: maybe only hide the second box (active) and change the "Inactive" label to simply "Label"

ghys replied:
Actually the active concept is still in effect because an item can still be defined when for "navigate" action types (if "show item value" is checked)...

LuckyMallari replied:
Implemented background color change on navigate since it still has to change color when item state changes.
Added active/inactive states to settings page.

ghys commented on web/app/widgets/button/button.widget.js:
```diff
@@ -103,45 +131,51 @@
             angular.extend(widget, $scope.form);
             switch (widget.action_type) {
                 case "navigate":
-                    delete widget.item;
```
Here's where you would do:
```
if ($scope.form.navigate_type) === 'dashboard') {
    delete widget.navigate_url;
    delete widget.navigate_target;
} else {
    delete widget.navigate_dashboard;
}
```

ghys commented on web/app/widgets/button/button.widget.js:
```diff
@@ -103,45 +131,51 @@
-            icon_replacestext: widget.icon_replacestext
+            icon_replacestext: widget.icon_replacestext,
+            show_item_value: widget.show_item_value || false,
+            navigate_type: widget.navigate_type || 'dashboard',
```
here you can derive the type as mentioned above:
```
                navigate_type: (widget.navigate_url) ? 'url' : 'dashboard',
```

ghys commented on web/app/widgets/button/button.tpl.html:
```diff
@@ -1,4 +1,11 @@
+        background-color: {{vm.background}};
+        color: {{vm.foreground}};
+    }
+</style>
```
Adding inline styles in `<style>`elements is discouraged unless in specific places like modals; furthermore, this won't work if you add several buttons - there is no concept of "scoped" or "virtual" DOM here so any added styling is global.

Styles for widgets should now be added in their associated .scss file if necessary, here the currently empty `button.scss`, and then processed with the `gulp sass` command.

I think the contents of the dummy widget template should rather be merged ("clerverly" or as a separate `<div ng-if="widget.show_item_value">`) into the button template instead of calling the `widget-dummy` directive. Of course the buttons should remain looking the same as before if the value is not shown.

LuckyMallari replied:
Yeah I thought about that also but since the button now can mimic the dummy, whenever there's any change in dummy will have to be done for the button template as well. With what I did, changes to dummy template will require no change in the button. 

Also if we want to mimic the dummy and still be able to change bg color, we'll have to implement that to the dummy itself. I can do that, let me know.

LuckyMallari replied:
Since this style is in that div, it will be scoped to that parent div. But with the new changes I will push, this will be irrelevant now. Im using the actual same html/tpl as dummy

ghys commented on web/app/widgets/button/button.widget.js:
```diff
@@ -57,17 +57,45 @@
+                return;
+
+            if (vm.widget.navigate_type === 'dashboard') {
+                $location.url('/view/' + vm.widget.navigate_url);
```
To ensure backwards compatibility the model property must remain `navigate_dashboard` - besides it's the dashboard id, not really an url.

In fact, you could get rid of `navigate_type` altogether in the model (not the form object in the settings dialog's scope) and simply derive it by looking at which property is defined:
```
if (vm.widget.navigate_dashboard) {
     // navigate to the dashboard in vm.widget.navigate_dashboard
} else {
     // navigate to the url in vm.widget.navigate_url
}
```
When submitting the settings, the property that isn't used (`navigate_dashboard` or `navigate_url` + `navigate_target`) should be deleted from the model (see below).

ghys commented:
> Since this style is in that div, it will be scoped to that parent div.

Wrong.

On Sat, Dec 30, 2017 at 8:33 PM, Lucky <notifications@github.com> wrote:

> *@LuckyMallari* commented on this pull request.
> ------------------------------
>
> In web/app/widgets/button/button.tpl.html
> <https://github.com/openhab/org.openhab.ui.habpanel/pull/259#discussion_r159128012>
> :
>
> > @@ -1,4 +1,11 @@
> -<div class="box-content switch" ng-style="{ 'background-color': vm.background, color: vm.foreground }" ng-click="vm.sendCommand()">
> +<style>
> +    .button-dummy .box-content.dummy {
> +        background-color: {{vm.background}};
> +        color: {{vm.foreground}};
> +    }
> +</style>
>
> Since this style is in that div, it will be scoped to that parent div. But
> with the new changes I will push, this will be irrelevant now. Im using the
> actual same html/tpl as dummy
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/org.openhab.ui.habpanel/pull/259#discussion_r159128012>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6Usxdca-jELsGbhkQ7_jjvH5OeVpQqks5tFpALgaJpZM4ROyPM>
> .
>



-- 
Yannick Schaus <mailto:yannick@schaus.net>

LuckyMallari commented on web/app/widgets/button/button.settings.tpl.html:
```diff
@@ -7,157 +7,239 @@ <h3>Button Settings</h3>
+                                        <label class="control-label col-lg-3 col-md-3">Link Target</label>
+                                        <div class="col-lg-9 col-md-9">
+                                            <label class="radio-inline">
+                                                <input type="radio" ng-model="form.navigate_target" value="self" name="navigate_target" checked/> self
```
Used "Current tab" instead

LuckyMallari commented:
Done all the changes. Had to add an active/inactive state if type is navigate. Otherwise navigate action type will have no clue how to switch colors based on item state since navigate type does not send a command..

![smartbutton_active_inactive](https://user-images.githubusercontent.com/6183345/34457332-9ea835d6-ed62-11e7-81f6-308d78a8d8c1.gif)

LuckyMallari commented:
Adding changes details here also

Implemented changes requested
* Used dummy template itself inside button template.
* Changed ui-select back to select tag
* Changed self to Current tab
* Captilalized 'targets'
* Kept Active/Inactive available on Navigate type
* Removed inline-style and used dummy html itself
* Brought back navigate_dashboard and kept it for backwards compatibility.

* Implemented background color change on navigate since it still has to change color when item state changes.
* Added active/inactive states to settings page.

LuckyMallari commented on web/app/widgets/button/button.settings.tpl.html:
```diff
@@ -246,6 +246,16 @@ <h3>Button Settings</h3>
                             </div>
                         </div>
                     </div>
+                    <div ng-if="form.action_type=='navigate' && form.show_item_value">
```
If the actiontype is navigate, the controller logic needs to know when to switch background/foreground colors. This setting only applies to navigate with show item value.

ghys replied:
Damn, you're right :) Completely overlooked this...
Your solution was nice but there is no "inactive" value - it's simply anything different than active :)

ghys commented:
@LuckyMallari thanks for your changes!
Also thanks for your animated gifs, really helpful :)

Nonetheless I finally checked out your branch and made a few finishing touches - ended up reorgnizing the settings dialog box so it's much more logical now...
While I was at it I added the ability to control the value display's font size and color independently of the label.

I think the button looks good now, and is much more powerful than before, I tested it thoroughly I think, no regressions detected, so I'll go ahead and merge it.

Again many thanks for your work! 👍

ghys commented:
Huh, you didn't sign off your commits - no need to amend the commit messages now, I'll do it when squashing them, just add a comment below with:

Signed-off-by: Your real name &lt;your@email&gt;

Thanks.

LuckyMallari commented:
Signed-off-by: Lucky Mallari <luckymallari@gmail.com>

servim commented:
Hi! I'm not sure what I'm writing in the right place, sorry :)
The button is a powerful widget, I like it very much, I use it to navigate between panels, displaying the value of some element on it. The only thing I really miss is the dynamic icon for the selected item.
Example: on the button icon - network. If the network element is turned on, the icon is blue, if off, gray. We see the gray icon, there is no network, click on the widget and switch to the panel - the network where the timeline widget will show the history of the network element.
Now I have to have 2 widgets, the dummy shows the state of the element, the button switches the panel, and I would like two in one!
